rust/xain-coordinator/src/training.rs

```diff
@@ -0,0 +1,337 @@
+use std::{
+    any::Any,
+    collections::{BTreeMap, BTreeSet},
+    fmt,
+    time::Duration,
+};
+
+use crate::{Model, ModelDim};
+
+macro_rules! err {
+    ($($tt:tt)*) => {
+        ProtocolError { msg: format!($($tt)*) }
+    }
+}
+
+macro_rules! bail {
+    ($($tt:tt)*) => {
+        return Err(err!($($tt)*))
+    }
+}
+
+#[derive(Debug)]
+pub struct ProtocolError {
+    msg: String,
+}
+
+impl fmt::Display for ProtocolError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&self.msg, f)
+    }
+}
+
+impl std::error::Error for ProtocolError {}
+
+type ParticipantId = u32;
+
+#[derive(Debug)]
+pub struct FromParticipant<T> {
+    pub from: ParticipantId,
+    pub payload: T,
+}
+
+#[derive(Debug)]
+pub enum InMessage {
+    Joined(FromParticipant<()>),
+    Update(FromParticipant<Model>),
+    RoundTimeout,
+}
+
+impl InMessage {
+    pub fn new_joined(from: ParticipantId) -> InMessage {
+        InMessage::Joined(FromParticipant { from, payload: () })
+    }
+
+    pub fn new_update(from: ParticipantId, model: Model) -> InMessage {
+        InMessage::Update(FromParticipant { from, payload: model })
+    }
+
+    pub fn new_round_timeout() -> InMessage {
+        InMessage::RoundTimeout
+    }
+}
+
+#[derive(Debug)]
+pub struct OutMessage {
+    pub to: ParticipantId,
+    pub kind: OutMessageKind,
+}
+
+#[derive(Debug)]
+pub enum OutMessageKind {
```

Comment:
I'm unsure about the semantics of OutMessage. Are those all broadcasted to participants?

Reply:
So, `OutMessage` has a `to` field, so they are not broadcasted. However, in the actual protocol, at the moment, we indeed just send every message to every client (by just looping). Seems like it might make sense to support explicit broadcast message.